### PR TITLE
Add API to return automatic compaction config history

### PR DESCRIPTION
**Pull request by suneet-s**
### Description

Add a new API to return the history of changes to automatic compaction config history to make it easy for users to see what changes have been made to their auto-compaction config.

The API is scoped per dataSource to allow users to triage issues with an individual dataSource. The API responds with a list of configs when there is a change to either the settings that impact all auto-compaction configs on a cluster or the dataSource in question.

I had considered adding another API to expose the entire compaction config history as it was stored in the audit table, however I chose against doing this, as I didn't think users would want the ability to see this global view.

#### Release note
New: You can now get the history of changes to automatic compaction config for a dataSource via an API.

<hr>

##### Key changed/added classes in this PR
 * `DataSourceCompactionConfigHistory#add` - Ensures that only changes that affect a particular dataSource's auto-compaction config are added to the history
 * `CoordinatorCompactionConfigsResource#getCompactionConfigHistory` - the API endpoint

<hr>

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/server/coordinator/DataSourceCompactionConfigHistory.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  private final String dataSource;
+  private DatasourceCompactionConfigAuditEntry current;
+
+  public DataSourceCompactionConfigHistory(String dataSource) {
+    this.dataSource = dataSource;
+  }
+
+  public void add(CoordinatorCompactionConfig coordinatorCompactionConfig, AuditInfo auditInfo, DateTime auditTime) {
+    for (DataSourceCompactionConfig dataSourceCompactionConfig : coordinatorCompactionConfig.getCompactionConfigs()) {
+      if (dataSource.equals(dataSourceCompactionConfig.getDataSource())) {
```
Comment (suneet-s) [inline server/src/main/java/org/apache/druid/server/coordinator/DataSourceCompactionConfigHistory.java:45]: Need to handle the case where someone deletes the compaction config for a datasource

File: server/src/main/java/org/apache/druid/server/coordinator/DatasourceCompactionConfigAuditEntry.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.druid.audit.AuditInfo;
+import org.joda.time.DateTime;
+
+/**
+ * A DTO containing audit information for compaction config for a datasource.
+ */
+public class DatasourceCompactionConfigAuditEntry
```
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/coordinator/DatasourceCompactionConfigAuditEntry.java:31]: Nit: Rename to `DataSourceCompactionConfigAuditEntry` with a capitalized `S` in `DataSource`, similar to rest of the classes here.

**Review by kfaraz — APPROVED** [review]
This is very useful, @suneet-s !
+1 after CI passes

File: server/src/main/java/org/apache/druid/server/http/CoordinatorCompactionConfigsResource.java [hunk 3]
```diff
@@ -168,6 +190,51 @@ public Response getCompactionConfig(@PathParam("dataSource") String dataSource)
...
+    try {
+      List<AuditEntry> auditEntries;
+      if (theInterval == null && count != null) {
+        auditEntries = auditManager.fetchAuditHistory(
+            CoordinatorCompactionConfig.CONFIG_KEY,
+            CoordinatorCompactionConfig.CONFIG_KEY,
+            count
+        );
+      } else {
+        auditEntries = auditManager.fetchAuditHistory(
```
Comment (maytasm) [inline server/src/main/java/org/apache/druid/server/http/CoordinatorCompactionConfigsResource.java:212]: Can a user pass in both interval and count?

Comment (suneet-s) [reply server/src/main/java/org/apache/druid/server/http/CoordinatorCompactionConfigsResource.java:212]: I copied the pattern in `CoordinatorDynamicConfigsResource#getDatasourceRuleHistory` where it looks like we ignore the count silently if the interval is specified.

**Review by maytasm — APPROVED** [review]

**Pull request merged**